src/combinatorics/simple_graph/inc_matrix.lean

```diff
@@ -0,0 +1,392 @@
+equal to the degree of `i`.
+5. `oriented_inc_matrix_elem_squared` : `(N(o) i e) ^ 2 = M i e`, for any orientation `o`,
+vertex `i` and edge `e`.
+6. `oriented_inc_matrix_mul_of_adj` : for any adjacent vertices `i`, `j` and edge `e`,
+`N(o) i e * N(o) j e = if e = ⟦(i, j)⟧ then -1 else 0`.
+7. `oriented_inc_matrix_mul_non_adj` : for any non-adjacent distinct vertices `i`, `j` and
+edge `e`, `N(o) i e * N(o) j e = 0`.
+8. `vec_mul_oriented_inc_matrix` : `(xᵀ ⬝ N(o)) e = x o.head(e) - x o.tail(e).`
+
+## References
+
+<https://en.wikipedia.org/wiki/Orientation_(graph_theory)>
+-/
+
+open_locale big_operators matrix
+open finset matrix simple_graph sym2
+
+universes u v
+variables {R : Type u} {V : Type v}
+
+section to_move
+
+@[simp]
+lemma ite_mul_ite_zero_right
+  {R : Type*} [mul_zero_class R] {P Q : Prop} [decidable P] [decidable Q] (a b : R):
+  (ite P a 0) * (ite Q b 0) = ite (P ∧ Q) (a * b) 0 :=
+by simp [←ite_and]
+
+end to_move
+
+namespace simple_graph
+
+variables (G : simple_graph V)
+
+-- ### Helping lemmas for edges
+
+lemma edge_set_ne {u v : V} {e: G.edge_set} (h : ↑e = ⟦(u, v)⟧) : u ≠ v :=
+begin
+  apply G.ne_of_adj,
+  rw [←G.mem_edge_set, ←h],
+  exact e.property
+end
+
+lemma incidence_equiv {i : V} {e : G.edge_set} : ↑e ∈ G.incidence_set i ↔ i ∈ (↑e : sym2 V) :=
+by simp only [incidence_set, true_and, set.mem_sep_eq, subtype.coe_prop]
+
+lemma any_vertex_if_incidence_set {i u v : V} :
+  ⟦(u, v)⟧ ∈ G.incidence_set i → i = u ∨ i = v :=
+by simp only [←mem_iff, incidence_set, set.mem_sep_eq, and_imp, imp_self, implies_true_iff]
+
+lemma incidence_set_if_any_vertex {i u v : V} (h : ⟦(u, v)⟧ ∈ G.edge_set) :
+  i = u ∨ i = v → ⟦(u, v)⟧ ∈ G.incidence_set i :=
+by simp only [←mem_iff, incidence_set, set.mem_sep_eq, h, true_and, imp_self]
+
+lemma edge_in_two_incidence_sets {i j : V} {e : sym2 V} (H_ne : i ≠ j) :
+  e ∈ G.incidence_set i ∧ e ∈ G.incidence_set j → e = ⟦(i, j)⟧ :=
+begin
+  refine quotient.rec_on_subsingleton e (λ p, _),
+  rcases p with ⟨v, w⟩,
+  rintros ⟨⟨-, H_i⟩, ⟨-, H_j⟩⟩,
+  rw ←elems_iff_eq H_ne,
+  tauto,
+end
+
+lemma mem_incidence_sets_iff_eq {i j : V} {e : sym2 V} (h : G.adj i j) :
+  e ∈ G.incidence_set i ∧ e ∈ G.incidence_set j ↔ e = ⟦(i, j)⟧ :=
+begin
+ split,
+  { apply edge_in_two_incidence_sets _,
+  rintro rfl,
+  exact G.loopless i h },
+  { rintro rfl,
+  nth_rewrite 1 eq_swap,
+  simp only [h, G.sym h, mem_incidence_set, and_self] }
+end
+
+lemma adj_iff_exists_edge_coe {i j : V} : G.adj i j ↔ ∃ (e : G.edge_set), ↑e = ⟦(i, j)⟧ :=
+by simp only [mem_edge_set, exists_prop, set_coe.exists, exists_eq_right, subtype.coe_mk]
+
+section incidence
+
+variables [fintype V] [decidable_eq V] [decidable_rel G.adj] (R) [ring R]
+
+-- ## Incidence matrix M
+
+/-- `inc_matrix G R` is the matrix `M` such that `M i e = 1` if vertex `i` is an
+endpoint of the edge `e` in the simple graph `G`, otherwise `M i j = 0`. -/
+def inc_matrix : matrix V G.edge_set R
```

Comment:
Sorry to suggest re-architecting this in a fundamental way, but I suspect things would be a lot easier if it were
```lean
def inc_matrix : matrix V (sym2 V) R
| i e := if e ∈ G.incidence_set i then 1 else 0

@[simp]
lemma inc_matrix_apply {i : V} {e : sym2 V} :
  G.inc_matrix R i e = if e ∈ G.incidence_set i then 1 else 0 := rfl
```
The tradeoff is that this isn't exactly the same as the usual definition of an incidence matrix.  This boldness here might be worth it, though.

You could then modify the first theorem statement to
```lean
theorem adj_sum_of_mul_inc_one {i j : V} (H_adj : G.adj i j) :
  ∑ e in G.edge_finset, G.inc_matrix R i e * G.inc_matrix R j e = (1 : R)
```

Reply:
I am not very convinced I should work with the column domain of the incidence matrix as all the possible pairs of vertices of the graph. This is not correct according to the mathematical definition of the incidence matrix, and it is also weird for me to think of a graph with 100 vertices and 2 connecting edges as a matrix having 9900 columns, when it should have only 2. Using `sym2 V` creates sparce matrices (even more sparce than they generally are).

Reply:
I agree it's weird, but the way it seems to simplify everything is very appealing to me -- sometimes these kinds of simplifications justify breaking from traditional definitions.  Plus, all the "fake" columns are all-zero, so the results are unaffected by this generalization.

It's also easy to restrict the matrix to `G.edge_set` if there's an application where you need to do this.  The theorems should still be able to apply to the restricted version.

This is similar to the "junk data principle" in formalization.  For example, while preventing division by zero is more correct, in practice it's easier to allow division by zero and choosing a junk value that we all agree never to exercise.  We can agree not to supply edges outside the `G.edge_set`. 

Though even if we do, having an all-zero column for those edges has good mathematical properties.  For example, this theorem surprisingly doesn't actually use that `e` is in `G.edge_finset`!
```lean
theorem adj_sum_of_mul_inc_one {i j : V} (H_adj : G.adj i j) :
  ∑ e, G.inc_matrix R i e * G.inc_matrix R j e = (1 : R) :=
begin
  simp only [G.mem_incidence_sets_iff_eq H_adj, inc_matrix_apply, mul_one, sum_boole,
    ite_mul_ite_zero_right],
  rw ←nat.cast_one,
  congr,
  convert_to _ = finset.card {⟦(i, j)⟧},
  { rw finset.card_singleton },
  { apply congr_arg finset.card,
    ext e,
    simp only [mem_filter, and_iff_right_iff_imp, mem_edge_finset, mem_singleton],
    rintro rfl,
    exact finset.mem_univ _ },
end
```

Reply:
For the record I agree with Kyle's position here, ultimately it almost always turns out nicer to formalise if data like this has redundant information.